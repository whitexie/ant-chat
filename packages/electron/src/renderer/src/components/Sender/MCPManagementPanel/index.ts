import { MCPManagementPanel } from './MCPManagementPanel'

export default MCPManagementPanel
