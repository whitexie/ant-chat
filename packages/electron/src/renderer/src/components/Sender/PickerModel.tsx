import type { AllAvailableModelsSchema } from '@ant-chat/shared'
import { RightOutlined, SettingOutlined } from '@ant-design/icons'
import { useRequest } from 'ahooks'
import { Input, Popover } from 'antd'
import React from 'react'
import { dbApi } from '@/api/dbApi'
import { getProviderLogo } from '../Chat/providerLogo'

interface PickerModelProps {
  value: AllAvailableModelsSchema['models'][number] | null
  onChange?: (value: AllAvailableModelsSchema['models'][number]) => void
}

export function PickerModel({ value, onChange }: PickerModelProps) {
  const [openPopover, setOpenPopover] = React.useState(false)
  const [keyword, setKeyword] = React.useState('')
  const { data } = useRequest(dbApi.getAllAbvailableModels)

  const activeProviderServiceInfo = !value ? data?.[0] : data?.find(item => item.models.some(model => model.id === value.id))

  React.useEffect(() => {
    if (!value && activeProviderServiceInfo?.models.length) {
      onChange?.(activeProviderServiceInfo?.models[0])
    }
  })

  return (
    <Popover
      open={openPopover}
      arrow={false}
      placement="bottomLeft"
      trigger="click"
      styles={{
        body: {
          padding: 0,
        },
      }}
      content={(
        <div>
          <div className="p-2 pt-2">
            <Input
              size="small"
              placeholder="搜索模型"
              value={keyword}
              onChange={e => setKeyword(e.target.value)}
            />
          </div>

          <div className="mt-1 max-h-38 px-2 overflow-y-auto">
            {
              data?.length
                ? (
                    data.map(item => (
                      <div key={item.id}>
                        <div className="text-xs text-gray-500 mt-1">{item.name}</div>
                        <div>
                          {item.models.filter(model => model.name.includes(keyword)).map(model => (
                            <div
                              key={model.id}
                              className="flex items-center gap-1 p-2 text-xs cursor-pointer hover:bg-(--hover-bg-color) rounded-md"
                              onClick={() => {
                                onChange?.(model)
                                setKeyword('')
                                setOpenPopover(false)
                              }}
                            >
                              {renderProviderLogo(item.id)}
                              <span className="font-medium">
                                {model.name}
                              </span>
                            </div>
                          ))}
                        </div>
                        <div className="h-1"></div>
                      </div>
                    ))
                  )
                : (
                    <div className="text-xs text-center p-2">
                      暂无模型，请先启用AI服务商
                    </div>
                  )
            }
          </div>

        </div>
      )}
      onOpenChange={setOpenPopover}
    >
      <div
        className={`
          group grid grid-cols-[max-content_0fr] hover:grid-cols-[max-content_1fr]
          transition-all duration-300
          h-8 rounded-md border-1 border-solid border-(--border-color) cursor-pointer
        `}
      >
        <div className="flex items-center pl-2 gap-1 hover:bg-(--hover-bg-color)">
          {renderProviderLogo(activeProviderServiceInfo?.id || '')}
          <div className="flex items-center text-xs font-medium max-w-30 truncate">
            <span className="truncate">{value?.name}</span>
            <RightOutlined className="px-2" />
          </div>
        </div>
        <div className="overflow-hidden flex items-center justify-center group-hover">
          <SettingOutlined
            className="px-2 h-full hover:bg-(--hover-bg-color)"
            onClick={(e) => {
              e.preventDefault()
              e.stopPropagation()
              // TODO 温度、最大TOKEN等配置
            }}
          />
        </div>
      </div>
    </Popover>
  )
}

function renderProviderLogo(providerServiceId: string) {
  const Logo = getProviderLogo(providerServiceId)

  if (!Logo)
    return null

  return <span className="bg-white p-1 rounded-md"><Logo /></span>
}
