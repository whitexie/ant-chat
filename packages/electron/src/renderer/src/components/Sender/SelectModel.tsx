import type { AllAvailableModelsSchema } from '@ant-chat/shared'
import { CheckOutlined } from '@ant-design/icons'
import { Input } from 'antd'
import React from 'react'
import { getProviderLogo } from '../Chat/providerLogo'

export interface SelectModelProps {
  value: string
  onChange?: (value: AllAvailableModelsSchema['models'][number]) => void
  options?: AllAvailableModelsSchema[]
}

export function SelectModel({ value, onChange, options }: SelectModelProps) {
  const [keyword, setKeyword] = React.useState('')

  return (
    <div>
      <div className="p-2 pt-2">
        <Input
          size="small"
          placeholder="搜索模型"
          value={keyword}
          onChange={e => setKeyword(e.target.value)}
        />
      </div>

      <div className="mt-1 max-h-38 overflow-y-auto px-2">
        {
          options?.length
            ? (
                options.map(item => (
                  <div key={item.id}>
                    <div className="mt-1 text-xs text-gray-500">{item.name}</div>
                    <div>
                      {item.models.filter(model => model.name.includes(keyword)).map(model => (
                        <div
                          key={model.id}
                          className={`
                            flex cursor-pointer items-center justify-between rounded-md p-2 text-xs
                            hover:bg-(--hover-bg-color)
                          `}
                          onClick={() => {
                            onChange?.(model)
                            setKeyword('')
                          }}
                        >
                          <div className="flex items-center gap-1">
                            {renderProviderLogo(item.id)}
                            <span className="font-medium">
                              {model.name}
                            </span>
                          </div>
                          <div>
                            {value === model.id ? (<CheckOutlined />) : null}
                          </div>
                        </div>
                      ))}
                    </div>
                    <div className="h-1"></div>
                  </div>
                ))
              )
            : (
                <div className="p-2 text-center text-xs">
                  暂无模型，请先启用AI服务商
                </div>
              )
        }
      </div>

    </div>
  )
}

export function renderProviderLogo(providerServiceId: string) {
  const Logo = getProviderLogo(providerServiceId)

  if (!Logo)
    return null

  return <span className="rounded-md bg-white p-1"><Logo /></span>
}
