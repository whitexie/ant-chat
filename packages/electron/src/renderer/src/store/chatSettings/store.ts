import type { AllAvailableModelsSchema } from '@ant-chat/shared'
import { create } from 'zustand'
import { devtools, persist } from 'zustand/middleware'

const initializeState = {
  model: null as AllAvailableModelsSchema['models'][number] | null,
  systemPrompt: '',
  temperature: 0.7,
  maxTokens: 4000,
}

export const useChatSttings = create<typeof initializeState>()(
  devtools(
    persist(
      () => ({ ...initializeState }),
      { name: 'chat-settings' },
    ),
    { enabled: import.meta.env.MODE === 'development' },
  ),
)
