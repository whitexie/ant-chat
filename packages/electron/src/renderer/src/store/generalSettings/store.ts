import type { GeneralSettingsState } from '@ant-chat/shared'
import { create } from 'zustand'
import { devtools } from 'zustand/middleware'

interface GeneralSettingsStoreState extends GeneralSettingsState {
  isLoading: boolean
}

const DEFAULT_SETTINGS: GeneralSettingsState = {
  assistantModelId: '',
  proxySettings: {
    mode: 'none',
    customProxyUrl: '',
  },
}

export const useGeneralSettingsStore = create<GeneralSettingsStoreState>()(
  devtools(
    _set => ({
      // Initial state
      ...DEFAULT_SETTINGS,
      isLoading: false,
    }),
    {
      name: 'General-Settings',
      enabled: import.meta.env.MODE === 'development',
    },
  ),
)
