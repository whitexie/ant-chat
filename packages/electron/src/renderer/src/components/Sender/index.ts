import Sender from './Sender'

export default Sender
