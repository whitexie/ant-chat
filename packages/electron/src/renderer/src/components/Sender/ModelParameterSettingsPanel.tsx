import type { ServiceProviderModelsSchema } from '@ant-chat/shared'
import { Input, Slider } from 'antd'
import { useEffect, useState } from 'react'
import { dbApi } from '@/api/dbApi'
import PromptIcon from '@/assets/icons/prompt.svg?react'
import ReturnIcon from '@/assets/icons/return.svg?react'
import TemperatureIcon from '@/assets/icons/temperature.svg?react'
import { useChatSettingsContext } from '@/contexts/chatSettings'

export function ModelParameterSettingsPanel() {
  const { settings, updateSettings } = useChatSettingsContext()

  const [modelInfo, setModelInfo] = useState<ServiceProviderModelsSchema | null>(null)

  useEffect(() => {
    const fetchModelInfo = async () => {
      const info = await dbApi.getModelInfoById(settings.modelId)
      setModelInfo(info)
    }
    fetchModelInfo()
  }, [settings.modelId])

  return (
    <div className="w-80 p-2 px-4">
      <div className="text-sm text-gray-500 mb-2">
        模型设置
      </div>
      <div className="form">
        <FormItem label="系统提示词" icon={<PromptIcon />}>
          <Input.TextArea value={settings.systemPrompt} onChange={e => updateSettings({ systemPrompt: e.target.value })} />
        </FormItem>
        <FormItem label="temperature" icon={<TemperatureIcon />}>
          <CustomSlider
            min={0}
            max={2}
            step={0.1}
            value={settings.temperature}
            onChange={value => updateSettings({ temperature: value })}
          />
        </FormItem>
        <FormItem label="maxTokens" icon={<ReturnIcon />}>
          <CustomSlider
            defaultValue={modelInfo?.maxTokens ?? 4096}
            min={1000}
            max={modelInfo?.maxTokens ?? 8000}
            formatter={value => `${Math.floor((value ?? 0) / 1000)}k`}
            step={1000}
            value={settings.maxTokens}
            onChange={value => updateSettings({ maxTokens: value })}
          />
        </FormItem>
      </div>
    </div>
  )
}

export function FormItem({ icon, label, children }: { icon?: React.ReactNode, label: string, children?: React.ReactNode }) {
  return (
    <div className="form-item py-2">
      <div className="form-item-label flex items-center gap-1 text-sm py-1">
        {icon}
        {label}
      </div>
      <div className="form-item-content">
        {children}
      </div>
    </div>
  )
}

interface CustomSliderProps {
  defaultValue?: number
  min: number
  max: number
  step: number
  value: number
  onChange: (value: number) => void
  formatter?: (value: number) => string
}

export function CustomSlider({ defaultValue, min, max, step, value, onChange, formatter }: CustomSliderProps) {
  return (
    <div className="relative">
      <div className="absolute top-0 right-0 translate-y-[-100%] text-xs">
        {formatter ? formatter(value) : value}
      </div>
      <Slider
        defaultValue={defaultValue}
        min={min}
        max={max}
        step={step}
        value={value}
        tooltip={{ open: false }}
        onChange={onChange}
      />
    </div>
  )
}
