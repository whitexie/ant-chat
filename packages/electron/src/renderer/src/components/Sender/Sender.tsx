import type { IAttachment, IImage } from '@ant-chat/shared'
import type { UploadFile } from 'antd'
import type { ChatFeatures } from '@/services-provider/interface'
import Icon, {
  ArrowUpOutlined,
  AudioOutlined,
  CloudUploadOutlined,
  GlobalOutlined,
  LinkOutlined,
} from '@ant-design/icons'
import { Attachments } from '@ant-design/x'
import useSpeech from '@ant-design/x/es/sender/useSpeech'
import { App, Badge, Button, Tooltip } from 'antd'
import { useState } from 'react'
import MCPIcon from '@/assets/icons/mcp.svg?react'
import StopSvg from '@/assets/icons/stop.svg?react'
import { useFeatures, useFeaturesState } from '@/store/features'
import { useMessagesStore } from '@/store/messages'
import { fileToBase64 } from '@/utils'
import SwitchButton from '../SwitchButton'

interface SenderProps {
  loading?: boolean
  actions?: React.ReactNode
  onSubmit?: (message: string, images: IImage[], attachments: IAttachment[], features: ChatFeatures) => void
  onCancel?: () => void
}

function Sender({ loading = false, actions, ...props }: SenderProps) {
  const { message } = App.useApp()
  const [text, setText] = useState('')
  const [openHeader, setOpenHeader] = useState(false)
  const [attachmentList, setAttachmentList] = useState<UploadFile[]>([])
  const [rows, setRows] = useState(2)
  const hasMessage = useMessagesStore(state => !!state.messages.length)
  const features = useFeaturesState()
  const { setOnlieSearch, setEnableMCP } = useFeatures()
  // 新增输入法状态
  const [isComposing, setIsComposing] = useState(false)

  const [allowSpeech, triggerSpeech, recording] = useSpeech((transcript: string) => {
    setText(prev => `${prev}${transcript}`)
  })

  async function transformAttachments() {
    const images: IAttachment[] = []
    const attachments: IAttachment[] = []

    if (attachmentList.length) {
      for (const item of attachmentList) {
        const { originFileObj, uid, name, type, size } = item as Required<UploadFile>
        const data = await fileToBase64(originFileObj)
        const result = { uid, name, size, type, data }
        if (type.includes('image')) {
          images.push(result)
        }
        else {
          attachments.push(result)
        }
      }
    }

    return { images, attachments }
  }

  async function handleSubmit() {
    const { images, attachments } = await transformAttachments()
    props?.onSubmit?.(text, images, attachments, features)
    setAttachmentList([])
    setText('')
    setOpenHeader(false)
    setRows(2)
  }

  return (
    <div
      className={`
        w-(--chat-width) mx-auto
        ${hasMessage ? 'relative' : 'absolute left-2 right-2 bottom-[50%]'} 
        transition transition-duration-500 p-3 
        bg-white dark:bg-[var(--ant-layout-color-bg-body)]
        border-solid border-2 border-[#0000001a] dark:border-[#fff6]
        rounded-xl overflow-hidden shadow-lg
        focus-within:border-(--ant-color-primary) focus-within:shadow-(--ant-box-shadow-secondary)
      `}
    >
      <div className={`header transition-height overflow-hidden h-0 ${openHeader && 'h-25'}`}>
        {
          openHeader && (
            <div data-testid="header-content" className="content">
              <Attachments
                multiple
                overflow="scrollX"
                beforeUpload={() => false}
                items={attachmentList}
                onChange={({ fileList }) => {
                  const result: UploadFile[] = []
                  fileList.forEach((item) => {
                    if (item.size && item.size > 1024 * 1024 * 20) {
                      message.warning('文件大小不能超过20MB')
                      return
                    }

                    // 识别Markdown文件
                    if (item.name.toLowerCase().endsWith('.md')) {
                      item.type = 'text/md'
                    }

                    result.push(item)
                  })

                  setAttachmentList(result)
                }}
                placeholder={{
                  icon: <CloudUploadOutlined />,
                  title: '上传图片或文档',
                }}
                accept="image/*,application/pdf,text/*,.md,.mp4"
              />
            </div>
          )
        }
      </div>
      <div className="input-wrapper">
        <textarea
          data-testid="textarea"
          value={text}
          rows={rows}
          onCompositionStart={() => setIsComposing(true)} // 输入法开始
          onCompositionEnd={() => setIsComposing(false)} // 输入法结束
          onChange={(e) => {
            const value = e.target.value
            setText(value)
            const length = value.split('\n').length
            if (length > 2 || rows > 2) {
              setRows(length < 6 ? length : 6)
            }
          }}
          onKeyDown={async (e) => {
            if (e.key === 'Enter' && !e.shiftKey && text.length > 0 && !isComposing) {
              e.preventDefault()
              await handleSubmit()
            }
          }}
          placeholder="Enter发送消息，Shift+Enter换行"
          className={`
            w-full h-full bg-transparent border-none p-1 outline-none resize-none 
            placeholder-[#b4b4b4] 
            dark:text-[var(--ant-layout-color-text-body)]  dark:placeholder-[var(--ant-layout-color-text-body)]`}
        />
      </div>
      <div className="footer flex justify-between">
        <div className="flex gap-1">
          <Tooltip title="附件(支持文档与图片)">
            <Badge dot={(attachmentList.length > 0) && !openHeader}>
              <Button
                data-testid="toggle-header"
                onClick={() => {
                  setOpenHeader(!openHeader)
                }}
                icon={<LinkOutlined />}
              />
            </Badge>
          </Tooltip>
          <Tooltip title="联网搜索(目前仅Gemini支持)">
            <div>
              <SwitchButton
                dataTestId="onlineSearch"
                checked={features.onlineSearch}
                onChange={setOnlieSearch}
                icon={<GlobalOutlined />}
              />
            </div>
          </Tooltip>
          <Tooltip title="MCP(token消耗较大)">
            <div>
              <SwitchButton
                checked={features.enableMCP}
                onChange={setEnableMCP}
                icon={<Icon component={MCPIcon} />}
              />
            </div>
          </Tooltip>
          {actions}
        </div>
        <div className="flex gap-1 items-center">
          <Button
            disabled={!allowSpeech}
            type="text"
            shape="circle"
            icon={recording ? <StopSvg className="w-8 h-8 color-[var(--ant-color-primary)]!" /> : <AudioOutlined />}
            onClick={() => {
              if (allowSpeech) {
                triggerSpeech(recording)
              }
            }}
          />
          <Button
            data-testid="sendBtn"
            type={loading ? 'text' : 'primary'}
            shape={loading ? 'default' : 'circle'}
            disabled={text.length === 0 && !loading}
            icon={loading
              ? (
                  <StopSvg className="w-8 h-8 !text-(--ant-color-primary)" />
                )
              : <ArrowUpOutlined />}
            onClick={
              () => {
                if (loading) {
                  props.onCancel?.()
                  return
                }
                handleSubmit()
              }
            }
          />
        </div>
      </div>
    </div>
  )
}

// function checkModel(model: AllAvailableModelsSchema['models'][number] | null) {
//   return !!model
// }

// function checkProviderServiceConfig(provider: AllAvailableModelsSchema) {
//   if (!provider.apiKey) {
//     return [false, 'apiKey 未配置']
//   }

//   if (!provider.baseUrl) {
//     return [false, 'baseUrl 未配置']
//   }

//   if (!provider.models.length) {
//     return [false, '该提供商下没有可用的模型']
//   }

//   return [true, '']
// }

export default Sender
