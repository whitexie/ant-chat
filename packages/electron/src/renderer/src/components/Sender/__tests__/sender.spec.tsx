import { checkModelConfig } from '@/store/modelConfig'
import { act, fireEvent, render, screen, waitFor } from '@testing-library/react'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import Sender from '../Sender'
import '@testing-library/jest-dom'

vi.mock('@/store/modelConfig')

describe('sender', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  it('loading状态下展示stop按钮', () => {
    render(
      <Sender loading />,
    )

    expect(screen.getByTitle('取消')).toBeInTheDocument()
  })

  describe('文本输入框', () => {
    it('按回车触发onSubmit', async () => {
      vi.mocked(checkModelConfig).mockReturnValue({ ok: true })

      const onSubmit = vi.fn()
      render(<Sender onSubmit={onSubmit} />)

      const input = screen.getByTestId('textarea')
      await act(async () => {
        fireEvent.change(input, { target: { value: '123\n123\n123' } })
      })

      await act(async () => {
        fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' })
      })

      expect(onSubmit).toHaveBeenCalledTimes(1)
      expect(onSubmit).toHaveBeenCalledWith('123\n123\n123', [], [], { enableMCP: false, onlineSearch: false })
    })

    it('文本为空时，按回车不触发onSubmit', async () => {
      const onSubmit = vi.fn()
      render(<Sender onSubmit={onSubmit} />)

      const input = screen.getByTestId('textarea')
      await act(async () => {
        fireEvent.change(input, { target: { value: '' } })
      })

      await act(async () => {
        fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' })
      })

      expect(onSubmit).toHaveBeenCalledTimes(0)
    })

    it('shift+Enter触发换行', async () => {
      const onSubmit = vi.fn()
      render(<Sender onSubmit={onSubmit} />)

      const input = screen.getByTestId('textarea')
      await act(async () => {
        fireEvent.change(input, { target: { value: '123' } })
      })

      await act(async () => {
        fireEvent.focus(input)
        fireEvent.keyDown(input, { key: 'Enter', code: 'Enter', shiftKey: true })
      })

      await waitFor(() => {
        expect(onSubmit).toHaveBeenCalledTimes(0)
      })

      expect((screen.getByTestId('textarea') as HTMLTextAreaElement).value).toBe('123')
    })
  })

  it('onCancel', () => {
    const onCancel = vi.fn()
    render(<Sender loading onCancel={onCancel} />)

    act(() => {
      fireEvent.click(screen.getByTitle('取消'))
    })

    expect(onCancel).toHaveBeenCalledTimes(1)
  })

  it('open header', async () => {
    render(<Sender />)

    // 找到触发打开头部的按钮
    const toggleButton = screen.getByTestId('toggle-header')

    // 点击按钮以打开头部
    await act(async () => {
      fireEvent.click(toggleButton)
    })

    const header = screen.getByTestId('header-content')
    expect(header).toBeInTheDocument()

    // 再次点击按钮以关闭头部
    await act(async () => {
      fireEvent.click(toggleButton)
    })

    // 验证头部的状态是否已恢复
    expect(header).not.toBeInTheDocument() // 验证标题是否不存在，表示头部已关闭
  })

  describe('发送按钮', () => {
    it('upload files', async () => {
      window.URL.createObjectURL = vi.fn(() => 'mocked-url')
      const onSubmit = vi.fn()

      render(<Sender onSubmit={onSubmit} />)

      // 点击按钮以打开头部
      await act(async () => {
        fireEvent.click(screen.getByTestId('toggle-header'))
      })

      // 模拟文件上传
      const file = new File(['file content'], 'example.md', { type: '' })
      const imgFile = new File([], 'image.png', { type: 'image/png' })
      const attachmentsComponent = screen.getByTestId('header-content')!

      // 触发 onChange 事件，模拟文件上传
      await act(async () => {
        fireEvent.change(attachmentsComponent.querySelector('input[type="file"]')!, {
          target: { files: [file, imgFile] },
        })
      })

      expect(screen.getByText('example')).toBeInTheDocument()

      await act(async () => {
        fireEvent.change(screen.getByTestId('textarea'), { target: { value: '123' } })
      })

      await act(async () => {
        fireEvent.click(screen.getByTestId('sendBtn'))
      })

      // 使用 waitFor 等待 onSubmit 被调用
      await waitFor(() => {
        expect(onSubmit).toHaveBeenCalledTimes(1)
      })

      expect(onSubmit).toHaveBeenCalledWith(
        '123',
        [expect.anything()],
        [expect.anything()],
        { enableMCP: false, onlineSearch: false },
      )
    })

    it('激活联网搜索', async () => {
      const onSubmit = vi.fn()
      render(<Sender onSubmit={onSubmit} />)

      vi.mocked(checkModelConfig).mockReturnValue({ ok: true })

      const onlineSearchBtn = screen.getByTestId('onlineSearch')
      const input = screen.getByTestId('textarea')
      const sendBtn = screen.getByTestId('sendBtn')

      await act(async () => {
        fireEvent.click(onlineSearchBtn)
      })

      await act(async () => {
        fireEvent.change(input, { target: { value: '123' } })
      })

      await act(async () => {
        fireEvent.click(sendBtn)
      })

      await waitFor(() => {
        expect(onSubmit).toHaveBeenCalledTimes(1)
      })

      expect(onSubmit).toHaveBeenCalledWith(
        '123',
        [],
        [],
        { enableMCP: false, onlineSearch: true },
      )
    })
  })
})
