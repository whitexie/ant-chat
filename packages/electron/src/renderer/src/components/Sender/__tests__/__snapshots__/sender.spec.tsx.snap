// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sender > initstate component 1`] = `
<DocumentFragment>
  <div
    class="md:(w-[var(--chat-width)]) mx-auto _sender_cdc83e absolute left-2 right-2 bottom-50% transition transition-duration-500 p-xs bg-white dark:(bg-[var(--ant-layout-color-bg-body)] border-white/40) rounded-xl overflow-hidden shadow-lg"
  >
    <div
      class="header transition-height overflow-hidden h-0 false"
    />
    <div
      class="input-wrapper"
    >
      <textarea
        class="w-full h-full scrollbar bg-transparent border-none p-1 outline-none resize-none placeholder-[#b4b4b4] text-size-base dark:(text-[var(--ant-layout-color-text-body)] placeholder-[var(--ant-layout-color-text-body)])"
        data-testid="textarea"
        placeholder="Enter发送消息，Shift+Enter换行"
        rows="2"
      />
    </div>
    <div
      class="footer flex justify-between"
    >
      <div
        class="flex gap-1"
      >
        <span
          aria-describedby="test-id"
          class="ant-badge css-dev-only-do-not-override-1v613y0"
        >
          <button
            class="ant-btn css-dev-only-do-not-override-1v613y0 ant-btn-default ant-btn-color-default ant-btn-variant-outlined ant-btn-icon-only"
            data-testid="toggle-header"
            type="button"
          >
            <span
              class="ant-btn-icon"
            >
              <span
                aria-label="link"
                class="anticon anticon-link"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="link"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M574 665.4a8.03 8.03 0 00-11.3 0L446.5 781.6c-53.8 53.8-144.6 59.5-204 0-59.5-59.5-53.8-150.2 0-204l116.2-116.2c3.1-3.1 3.1-8.2 0-11.3l-39.8-39.8a8.03 8.03 0 00-11.3 0L191.4 526.5c-84.6 84.6-84.6 221.5 0 306s221.5 84.6 306 0l116.2-116.2c3.1-3.1 3.1-8.2 0-11.3L574 665.4zm258.6-474c-84.6-84.6-221.5-84.6-306 0L410.3 307.6a8.03 8.03 0 000 11.3l39.7 39.7c3.1 3.1 8.2 3.1 11.3 0l116.2-116.2c53.8-53.8 144.6-59.5 204 0 59.5 59.5 53.8 150.2 0 204L665.3 562.6a8.03 8.03 0 000 11.3l39.8 39.8c3.1 3.1 8.2 3.1 11.3 0l116.2-116.2c84.5-84.6 84.5-221.5 0-306.1zM610.1 372.3a8.03 8.03 0 00-11.3 0L372.3 598.7a8.03 8.03 0 000 11.3l39.6 39.6c3.1 3.1 8.2 3.1 11.3 0l226.4-226.4c3.1-3.1 3.1-8.2 0-11.3l-39.5-39.6z"
                  />
                </svg>
              </span>
            </span>
          </button>
        </span>
        <div
          aria-describedby="test-id"
        >
          <div
            class="ant-btn antd-css-var w-8 h-8 flex justify-center items-center border-1 border-solid cursor-pointer"
            role="switchButton"
            style="border-radius: 6px; color: var(--ant-button-default-color);"
          >
            <span
              aria-label="global"
              class="anticon anticon-global"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="global"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M854.4 800.9c.2-.3.5-.6.7-.9C920.6 722.1 960 621.7 960 512s-39.4-210.1-104.8-288c-.2-.3-.5-.5-.7-.8-1.1-1.3-2.1-2.5-3.2-3.7-.4-.5-.8-.9-1.2-1.4l-4.1-4.7-.1-.1c-1.5-1.7-3.1-3.4-4.6-5.1l-.1-.1c-3.2-3.4-6.4-6.8-9.7-10.1l-.1-.1-4.8-4.8-.3-.3c-1.5-1.5-3-2.9-4.5-4.3-.5-.5-1-1-1.6-1.5-1-1-2-1.9-3-2.8-.3-.3-.7-.6-1-1C736.4 109.2 629.5 64 512 64s-224.4 45.2-304.3 119.2c-.3.3-.7.6-1 1-1 .9-2 1.9-3 2.9-.5.5-1 1-1.6 1.5-1.5 1.4-3 2.9-4.5 4.3l-.3.3-4.8 4.8-.1.1c-3.3 3.3-6.5 6.7-9.7 10.1l-.1.1c-1.6 1.7-3.1 3.4-4.6 5.1l-.1.1c-1.4 1.5-2.8 3.1-4.1 4.7-.4.5-.8.9-1.2 1.4-1.1 1.2-2.1 2.5-3.2 3.7-.2.3-.5.5-.7.8C103.4 301.9 64 402.3 64 512s39.4 210.1 104.8 288c.2.3.5.6.7.9l3.1 3.7c.4.5.8.9 1.2 1.4l4.1 4.7c0 .1.1.1.1.2 1.5 1.7 3 3.4 4.6 5l.1.1c3.2 3.4 6.4 6.8 9.6 10.1l.1.1c1.6 1.6 3.1 3.2 4.7 4.7l.3.3c3.3 3.3 6.7 6.5 10.1 9.6 80.1 74 187 119.2 304.5 119.2s224.4-45.2 304.3-119.2a300 300 0 0010-9.6l.3-.3c1.6-1.6 3.2-3.1 4.7-4.7l.1-.1c3.3-3.3 6.5-6.7 9.6-10.1l.1-.1c1.5-1.7 3.1-3.3 4.6-5 0-.1.1-.1.1-.2 1.4-1.5 2.8-3.1 4.1-4.7.4-.5.8-.9 1.2-1.4a99 99 0 003.3-3.7zm4.1-142.6c-13.8 32.6-32 62.8-54.2 90.2a444.07 444.07 0 00-81.5-55.9c11.6-46.9 18.8-98.4 20.7-152.6H887c-3 40.9-12.6 80.6-28.5 118.3zM887 484H743.5c-1.9-54.2-9.1-105.7-20.7-152.6 29.3-15.6 56.6-34.4 81.5-55.9A373.86 373.86 0 01887 484zM658.3 165.5c39.7 16.8 75.8 40 107.6 69.2a394.72 394.72 0 01-59.4 41.8c-15.7-45-35.8-84.1-59.2-115.4 3.7 1.4 7.4 2.9 11 4.4zm-90.6 700.6c-9.2 7.2-18.4 12.7-27.7 16.4V697a389.1 389.1 0 01115.7 26.2c-8.3 24.6-17.9 47.3-29 67.8-17.4 32.4-37.8 58.3-59 75.1zm59-633.1c11 20.6 20.7 43.3 29 67.8A389.1 389.1 0 01540 327V141.6c9.2 3.7 18.5 9.1 27.7 16.4 21.2 16.7 41.6 42.6 59 75zM540 640.9V540h147.5c-1.6 44.2-7.1 87.1-16.3 127.8l-.3 1.2A445.02 445.02 0 00540 640.9zm0-156.9V383.1c45.8-2.8 89.8-12.5 130.9-28.1l.3 1.2c9.2 40.7 14.7 83.5 16.3 127.8H540zm-56 56v100.9c-45.8 2.8-89.8 12.5-130.9 28.1l-.3-1.2c-9.2-40.7-14.7-83.5-16.3-127.8H484zm-147.5-56c1.6-44.2 7.1-87.1 16.3-127.8l.3-1.2c41.1 15.6 85 25.3 130.9 28.1V484H336.5zM484 697v185.4c-9.2-3.7-18.5-9.1-27.7-16.4-21.2-16.7-41.7-42.7-59.1-75.1-11-20.6-20.7-43.3-29-67.8 37.2-14.6 75.9-23.3 115.8-26.1zm0-370a389.1 389.1 0 01-115.7-26.2c8.3-24.6 17.9-47.3 29-67.8 17.4-32.4 37.8-58.4 59.1-75.1 9.2-7.2 18.4-12.7 27.7-16.4V327zM365.7 165.5c3.7-1.5 7.3-3 11-4.4-23.4 31.3-43.5 70.4-59.2 115.4-21-12-40.9-26-59.4-41.8 31.8-29.2 67.9-52.4 107.6-69.2zM165.5 365.7c13.8-32.6 32-62.8 54.2-90.2 24.9 21.5 52.2 40.3 81.5 55.9-11.6 46.9-18.8 98.4-20.7 152.6H137c3-40.9 12.6-80.6 28.5-118.3zM137 540h143.5c1.9 54.2 9.1 105.7 20.7 152.6a444.07 444.07 0 00-81.5 55.9A373.86 373.86 0 01137 540zm228.7 318.5c-39.7-16.8-75.8-40-107.6-69.2 18.5-15.8 38.4-29.7 59.4-41.8 15.7 45 35.8 84.1 59.2 115.4-3.7-1.4-7.4-2.9-11-4.4zm292.6 0c-3.7 1.5-7.3 3-11 4.4 23.4-31.3 43.5-70.4 59.2-115.4 21 12 40.9 26 59.4 41.8a373.81 373.81 0 01-107.6 69.2z"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
      <div>
        <button
          class="ant-btn css-dev-only-do-not-override-1v613y0 ant-btn-circle ant-btn-primary ant-btn-color-primary ant-btn-variant-solid ant-btn-icon-only"
          data-testid="sendBtn"
          disabled=""
          type="button"
        >
          <span
            class="ant-btn-icon"
          >
            <span
              aria-label="arrow-up"
              class="anticon anticon-arrow-up"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="arrow-up"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M868 545.5L536.1 163a31.96 31.96 0 00-48.3 0L156 545.5a7.97 7.97 0 006 13.2h81c4.6 0 9-2 12.1-5.5L474 300.9V864c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V300.9l218.9 252.3c3 3.5 7.4 5.5 12.1 5.5h81c6.8 0 10.5-8 6-13.2z"
                />
              </svg>
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
